### PR TITLE
Lab 1077 max stable set

### PR DESCRIPTION
**Pull request by KasiaKoz**
This is an absolute chonker. So I do not expect anyone to review all or any of this, but I would appreciate it if you could have a glance as some bits and bobs or just note that this is making a move to master at long last.

This PR adds three major methods to relate schedule elements to the network.
1. `route_schedule` - snaps and routes multiple services or the whole schedule. Not recommended to do the whole schedule for big and complicated PT networks but this method is good for multiple services, because it buffers changes to network and the services until the end. Rather than looping through the services and doing it for single service each time as in the following method:
2. `route_service` - snaps and routes a service to the network
3. `teleport_service` - teleports a service, useful as a last resort or for modes that aren't available in the network (e.g. ferry)

[Here is a medium blog post](https://medium.com/arupcitymodelling/putting-transit-on-a-map-with-genet-11dd7bf835d7) that is a good intro to the work in this branch.

The two snapping methods use Max Stable Set graph optimisation problem to choose links in the network of the correct mode that are closest to the stops and one that you can find a route for. 

## Max Stable Set problem

V is the set of nodes, E is the set of edges; $x_i$ takes on the value 1 when node 𝑖 is included in the stable set, and 0 if not:
Screenshot 2021-10-28 at 11 28 46
i.e. Find the largest number of nodes such that none of them share an edge ([ref1](https://scipbook.readthedocs.io/en/latest/graph.html#maximum-stable-set-problem), [ref2](https://mathworld.wolfram.com/MaximumIndependentVertexSet.html)).

In our case, we also add a constant for each $x_i$ that is a proxy scalar value computed based on route distances from that node to all other possible nodes (that are to be connected on the PT route), so for us it's a sum of $C_i * x_i$. This constant is larger if the length of the route is smaller, making it a more desirable choice when maximising the sum.

We build an abstract graph that represents the problem of snapping GTFS to an existing network in terms of the max stable set formulation. The nodes represent the links in the network within a snapping distance threshold that a PT stop could snap to. These are completely connected within their catchments - by max stable set, we will only be able to select one of those nodes. We then also add edges between some nodes between these catchments if there isn't a route between those two links in the network - meaning we could select one of them, but not both because then we would not be able to find a route between them.

image

## Tested with Matesto ™️ 
attached a GTFS based schedule to a Puma network.xml (because it's a verified good network.xml file)

Screenshot 2021-10-26 at 16 40 55

Screenshot 2021-10-26 at 16 41 01

### CODE REVIEW

**Review by gac55 — APPROVED** [review]
Somewhat beyond my ability to review ;) 

The blog post is very helpful. The output results are intuitively looking sensible and it gives us something far more visibly appealing. 

A wider question - could we use some data from OSM to better inform this? I'm thinking of tangible tags that say if a bus is/isn't permitted on a link and softer things, such as small/narrow lanes where a bus is unlikely to be able to navigate. I know navigation systems do similar logic to at least ensure they don't travel on links they are unlikely to be able to pass (or even pass safely). 

These are likely edge case issues, but I am curious what the next iteration of improvements may look like

**Comment by gac55** [comment]
Also 👍 to see Matesto in use

File: tests/test_max_stable_set.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+                                      ])])
+    return n
+
+
+@pytest.fixture()
+def network_spatial_tree(network):
+    return spatial.SpatialTree(network)
+
+
+def test_all_stops_have_nearest_links_returns_False_with_missing_closest_links(mocker, network, network_spatial_tree):
```
Comment (mfitz) [inline tests/test_max_stable_set.py:111]: How about `test_detects_stops_that_lack_nearest_links` or something that reads more like a description of behaviour and is not coupled to a method name?

File: tests/test_max_stable_set.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+    mss.unsolved_stops = {'stop_1'}
+    return mss
+
+
+def test_generating_changeset_from_partial_mss_problem(partial_mss):
+    changeset = partial_mss.to_changeset(
+        DataFrame({'ordered_stops': {'service_1_route_2': ['stop_3', 'stop_2', 'stop_1'],
+                                     'service_1_route_1': ['stop_1', 'stop_2', 'stop_3']}})
+    )
+    assert_semantically_equal(
```
Comment (mfitz) [inline tests/test_max_stable_set.py:424]: Is there an argument for each of these assertions becoming its own test? Ideally we want to minimise the number of things that can make a test fail, and this test seems to have a high number of different things that could cause failure.

**Review by mfitz — APPROVED** [review]
I haven't really grokked the essence of the max stable set cleverness, but it looks good as far as I can see 👍

File: genet/max_stable_set.py [hunk 3]
```diff
@@ -0,0 +1,442 @@
...
+            if not isinstance(_out, int):
+                _out = 0
+            if not isinstance(_in, int):
+                _in = 0
+            return _in + _out
+
+    def has_a_completely_connected_catchment(self):
+        stop_id_groups = self.nodes.groupby('id')
+        for u, v in self.pt_graph.edges():
+            try:
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply genet/max_stable_set.py]: did a quick test with the unit tests, so not super serious.. but managed to shave off some secs
before
```
================== 11 passed, 2 warnings in 79.07s (0:01:19) ===================...........
```
after
```
================== 11 passed, 2 warnings in 72.37s (0:01:12) ===================...........
```

File: genet/outputs_handler/geojson.py [hunk 4]
```diff
@@ -163,7 +165,7 @@ def generate_standard_outputs(n, output_dir, gtfs_day='19700101', include_shp_fi
 
     logging.info('Generating geojson outputs for car/driving modal subgraph')
     graph_output_dir = os.path.join(output_dir, 'graph')
-    gdf_car = graph_links[graph_links.apply(lambda x: modal_subset(x, {'car'}), axis=1)]
+    gdf_car = graph_links.loc[graph_links.apply(lambda x: modal_subset(x, {'car'}), axis=1), :]  # noqa: E231
```
[the remark this answers is not in the document]
Comment (KasiaKoz) [reply genet/outputs_handler/geojson.py:168]: so i'm in a catch 22
```
./genet/outputs_handler/geojson.py:168:94: E231 missing whitespace after ':'
    gdf_car = graph_links.loc[graph_links.apply(lambda x: modal_subset(x, {'car'}), axis=1), :]
                                                                                             ^
1     E231 missing whitespace after ':'
```
of if I add the white space `gdf_car = graph_links.loc[graph_links.apply(lambda x: modal_subset(x, {'car'}), axis=1), : ]` I get
```
./genet/outputs_handler/geojson.py:168:95: E202 whitespace before ']'
    gdf_car = graph_links.loc[graph_links.apply(lambda x: modal_subset(x, {'car'}), axis=1), : ]
                                                                                              ^
1     E202 whitespace before ']'
```
I think I'll just stay with switching it off XD
